node/src/components/consensus/protocols/zug.rs

```diff
@@ -2599,3 +2622,63 @@ mod specimen_support {
         }
     }
 }
+
+mod registered_sync {
+    use crate::{
+        types::{DataSize, NodeRng},
+        utils::specimen::{Cache, LargestSpecimen, SizeEstimator},
+    };
+    use casper_types::{TimeDiff, Timestamp};
+    use rand::Rng as _;
+    use serde::{Deserialize, Serialize};
+    use std::collections::BTreeMap;
+
+    #[derive(Default, DataSize, Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Hash)]
+    pub struct RegisteredSync(BTreeMap<RandomId, Timestamp>);
+
+    #[derive(
+        DataSize, Copy, Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Hash, Ord, PartialOrd,
+    )]
+    pub struct RandomId(u64);
+
+    impl RegisteredSync {
+        /// Prunes entries older than one minute.
+        fn prune_old(&mut self) {
+            const ONE_MIN: TimeDiff = TimeDiff::from_seconds(60);
+
+            self.0.retain(|_, timestamp| timestamp.elapsed() < ONE_MIN);
+        }
+
+        pub fn insert_new(&mut self, rng: &mut NodeRng) -> RandomId {
+            self.prune_old();
+
+            let id = loop {
+                let id = RandomId::new(rng);
+
+                if self.0.contains_key(&id) == false {
+                    break id;
+                }
+            };
+
+            self.0.insert(id, Timestamp::now());
+
+            id
+        }
+
+        pub fn id_not_registered(&mut self, id: RandomId) -> bool {
```

Comment:
I'd call this something like `try_remove_id`. `id_not_registered` sounds like it only checks for the ID having been registered, but it actually has a side effect of temoving it.

Reply:
Fair point.